ui/frontend/selectors/index.ts

```diff
@@ -241,6 +242,16 @@ export const issueUrlSelector = createSelector(
   },
 );
 
+export const codeUrlSelector = createSelector(
+  codeSelector,
+  code => {
+    const codeUrl =  url.parse(window.location.href, true);
+    codeUrl.query.code = code;
+    codeUrl.search = qs.stringify(codeUrl.query.code);
```

Comment:
Why are you setting both `query` and `search`? Check out the previous `*UrlSelector` functions which just set the `query` parameter and rely on the url library to properly escape the value. Then we don't need to import `qs` at all.

Reply:
It works when no any other query presents. However, if you switch channels or alter other configs that add query strings, `url.format` would ignore your query object and use `urlObj.search` directly. See [`url.format(urlObj)`](https://www.npmjs.com/package/url/v/0.11.0#urlformaturlobj).

I know it is a little hacky here. Here is an alternative approach using `URLSearchParams` instead of `url` module. The drawback is that it does not support IE or older browsers.

```rust
code => {
    const codeUrl = new URL(window.location.href);
    codeUrl.searchParams.append('code', code);
    return codeUrl.href;
  },
```

Reply:
> switch channels or alter other configs that add query strings

Then let's reuse the code that generates those query strings (`permalinkSelector`).

Something like:

```js
export const xxxSelector = createSelector(
  baseUrlSelector, gistSelector,
  (baseUrl, gist) => {
    const u = url.parse(baseUrl, true);
    u.query = {
      version: gist.channel,
      mode: gist.mode,
      edition: gist.edition,
    };
    return u;
  },
);

export const permalinkSelector = createSelector(
  xxxSelector, gistSelector,
  (u, gist) => {
    u.query.gist = gist.id;
    return url.format(u);
  },
);

export const codeUrlSelector = createSelector(
  xxxSelector, codeSelector,
  (u, code) => {
    u.query.code = code;
    return url.format(u);
  },
);
```

Reply:
The example above cannot work. It is a misuse of selectors. Selectors memoize data selected and continue passing by reference. Modify argument would influence all selectors using it. 

Back to my modified code similar to this example, the **permalink** link would get a `code=fn main()...` query string.

Reply:
> , a selector must be a pure function without any modification of arguments.

Fair enough, but the core idea is still valid:

```js
export const xxxSelector = createSelector(
  gistSelector,
  (gist) => ({
    version: gist.channel,
    mode: gist.mode,
    edition: gist.edition,
  }),
);

export const permalinkSelector = createSelector(
  baseUrlSelector, xxxSelector, gistSelector,
  (baseUrl, xxx, gist) => {
    const u = url.parse(baseUrl, true);
    u.query = { gist: gist.id, ...xxx };
    return url.format(u);
  },
);

export const codeUrlSelector = createSelector(
  baseUrlSelector, xxxSelector, codeSelector,
  (baseUrl, xxx, code) => {
    const u = url.parse(baseUrl, true);
    u.query = { code, ...xxx };
    return url.format(u);
  },
);
```

Reply:
Modified at 3479820.